controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,202 @@
+package storageautoscaler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"errors"
+
+	"github.com/go-logr/logr"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StorageAutoscalerReconciler is the reconciler for StorageAutoscaler objects
+type StorageAutoscalerReconciler struct {
+	client.Client
+	Log               logr.Logger
+	OperatorNamespace string
+	SyncMap           *sync.Map
+	EventCh           chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// get the eventCh from the storage autoscaler scraper
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.EventCh,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=*,verbs=get
+
+// Reconcile reconciles the StorageAutoscaler object
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.Log.Info("reconciling storage autoscaler", "namespace", request.Namespace, "name", request.Name)
+
+	// list the storage autoscaler reconciler object
+	storageAutoScaler := &ocsv1.StorageAutoScaler{}
+	err := r.Get(ctx, client.ObjectKey{Namespace: request.Namespace, Name: request.Name}, storageAutoScaler)
+	if err != nil {
+		r.Log.Error(err, "failed to get storage autoscaler")
+		return reconcile.Result{}, err
+	}
+
+	// if no expansion is in-progress update the phase to "not-started"
+	if storageAutoScaler.Status.Phase == "" {
+		storageAutoScaler.Status.Phase = "NotStarted"
+		err := r.updateStatus(ctx, storageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if the expansion is in-progress, verify the scaling
+	if storageAutoScaler.Status.Phase == "InProgress" {
+		return r.verifyScaling(ctx, storageAutoScaler)
+	}
+
```

Comment:
@travisn 

how about checking if the `since(last-expansion.completionTime ) > 10 min`? 

So we would be sure the sync map is updated for the new resize?

Reply:
Why the delay? If the previous expansion is in `Succeeded` state, wouldn't the sync map already be updated? Are you concerned about a new expansion being triggered too soon after a horizontal scale when we just need to wait for the PGs to be rebalanced? Instead of some wait like that, it will be much more reliable if we only scale when the PGs are clean.

Reply:
Yes for sure we can check if the cephcluster health is ok, that would definitely help.

But this will actually reject the cache reconcile events that were triggered before the reconcile get to succeeded

Reply:
For pgs health, should we check cephcluster status, or query prometheus?

Reply:
Per discussion, either cephcluster status or prometheus query will be valid, although prometheus query may be more updated. The cephcluster status could be up to a minute outdated. Per discussion, anyway let's go with the prometheus query.